rasa/core/training/story_conflict.py

```
@@ -302,19 +302,38 @@ def _get_previous_event(
+            if (
+                "action_name" in substate
+                and substate["action_name"] != ACTION_LISTEN_NAME
+            ):
+                # The `prev_...` was an action that was NOT `action_listen`
+                if not isinstance(substate["action_name"], str):
+                    # While the Substate type doesn't restrict the value of `action_name` to be a string, it always should be
+                    raise TypeError(
+                        f"The value '{substate['action_name']}' of `action_name` should be a string, not {type(substate['action_name'])}. Did you modify Rasa source code?"
+                    )
+                return "action", substate["action_name"]
+            elif "action_text" in substate:
+                # The `prev_...` was a a free form utterance action
+                if not isinstance(substate["action_text"], str):
+                    # While the Substate type doesn't restrict the value of `action_text` to be a string, it always should be
+                    raise TypeError(
+                        f"The value '{substate['action_text']}' of `action_text` should be a string, not {type(substate['action_text'])}. Did you modify Rasa source code?"
+                    )
+                return "bot uttered", substate["action_text"]
+        elif origin == USER:
             # We found an intent, but it is only the previous event if
             # the `prev_...` was `prev_action_listen`, so we don't return.
-            previous_event_type = "intent"
-            previous_event_name = turn_label.replace(INTENT + "_", "")
-
+            if "intent" in substate:
+                previous_event_type = "intent"
+                previous_event_name = substate["intent"]
+            elif "text" in substate:
+                previous_event_type = "user utterance"
+                previous_event_name = substate["text"]
+
+    assert isinstance(previous_event_name, (str, type(None)))
```

I guess we should raise the same exception here.

Looking at the code again, you could also the logic of 
```
                previous_event_type = "intent"
                previous_event_name = substate["intent"]
```
in the if branch `if origin == PREVIOUS_ACTION:`. Then you would just need to raise the exception once.

---

Ah, forgot about that. :man_facepalming: 
Actually... now that states are dicts we don't even need the loop anymore I think...